bindings/postmark/postmark.go

```diff
@@ -0,0 +1,154 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package postmark
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/dapr/pkg/logger"
+	"github.com/keighl/postmark"
+)
+
+// Postmark allows sending of emails using the 3rd party Postmark service
+type Postmark struct {
+	metadata postmarkMetadata
+	logger   logger.Logger
+}
+
+// Our metadata holds standard email properties
+type postmarkMetadata struct {
+	ServerToken  string `json:"serverToken"`
+	AccountToken string `json:"accountToken"`
+	EmailFrom    string `json:"emailFrom"`
+	EmailTo      string `json:"emailTo"`
+	Subject      string `json:"subject"`
+	EmailCc      string `json:"emailCc"`
+	EmailBcc     string `json:"emailBcc"`
+}
+
+// NewPostmark returns a new Postmark bindings instance
+func NewPostmark(logger logger.Logger) *Postmark {
+	return &Postmark{logger: logger}
+}
+
+// Helper to parse metadata
+func (p *Postmark) parseMetadata(meta bindings.Metadata) (postmarkMetadata, error) {
+	pMeta := postmarkMetadata{}
+
+	// Required properties
+	if val, ok := meta.Properties["serverToken"]; ok && val != "" {
+		pMeta.ServerToken = val
+	} else {
+		return pMeta, errors.New("Postmark binding error: serverToken field is required in metadata")
+	}
+	if val, ok := meta.Properties["accountToken"]; ok && val != "" {
+		pMeta.AccountToken = val
+	} else {
+		return pMeta, errors.New("Postmark binding error: accountToken field is required in metadata")
+	}
+
+	// Optional properties, these can be set on a per request basis
+	pMeta.EmailTo = meta.Properties["emailTo"]
+	pMeta.EmailFrom = meta.Properties["emailFrom"]
+	pMeta.Subject = meta.Properties["subject"]
+	pMeta.EmailCc = meta.Properties["emailCc"]
+	pMeta.EmailBcc = meta.Properties["emailBcc"]
+
+	return pMeta, nil
+}
+
+// Init does metadata parsing and not much else :)
+func (p *Postmark) Init(metadata bindings.Metadata) error {
+	// Parse input metadata
+	meta, err := p.parseMetadata(metadata)
+	if err != nil {
+		return err
+	}
+
+	// Um, yeah that's about it!
+	p.metadata = meta
+
+	return nil
+}
+
+func (p *Postmark) Operations() []bindings.OperationKind {
+	return []bindings.OperationKind{bindings.CreateOperation}
+}
+
+// Write does the work of sending message to Postmark API
+func (p *Postmark) Invoke(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	// We allow two possible sources of the properties we need,
+	// the component metadata or request metadata, request takes priority if present
+
+	var email postmark.Email
+
+	// Build email from address, this is required
+	if p.metadata.EmailFrom != "" {
```

Comment:
It's best to move the logic of required fields validation to Init and fail once on startups rather than for every operation.

Reply:
But the logic a fall-trough, the initial values are optional and can be overwritten by the request, so they can't be validated on init, only on the actual request.

Reply:
ok great